pkg/util/encryption/multi_test.go

```diff
@@ -0,0 +1,156 @@
+			},
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			controller := gomock.NewController(t)
+			defer controller.Finish()
+
+			kv := mock_keyvault.NewMockManager(controller)
+			tt.mocks(tt, kv)
+
+			_, err := NewMulti(ctx, kv, tt.secretName, tt.legacySecretName)
+			if err != nil && err.Error() != tt.wantErr ||
+				err == nil && tt.wantErr != "" {
+				t.Fatal(err)
+			}
+		})
+	}
+}
+
+func TestOpen(t *testing.T) {
+	ctx := context.Background()
+
+	newKey := []byte("\x6a\x98\x95\x6b\x2b\xb2\x7e\xfd\x1b\x68\xdf\x5c\x40\xc3\x4f\x8b\xcf\xff\xe8\x17\xc2\x2d\xf6\x40\x2e\x5a\xb0\x15\x63\x4a\x2d\x2e\xab\x79\x86\x50\xfb\xce\xdc\x9d\xdd\x1c\x01\x32\xd6\x03\x99\xe6\x59\x81\x37\xb3\xdb\x67\x6f\x12\x34\x1d\xb9\x58\x18\x31\x30\x57")
+	newOpener, err := NewAES256SHA512(ctx, newKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	legacyKey := []byte("\x6a\x98\x95\x6b\x2b\xb2\x7e\xfd\x1b\x68\xdf\x5c\x40\xc3\x4f\x8b\xcf\xff\xe8\x17\xc2\x2d\xf6\x40\x2e\x5a\xb0\x15\x63\x4a\x2d\x2e")
+	legacyOpener, err := NewXChaCha20Poly1305(ctx, legacyKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	multi := multi{
+		openers: []AEAD{
+			newOpener,
+			legacyOpener,
+		},
+	}
+
+	type test struct {
+		name    string
+		input   []byte
+		wantErr string
+	}
+
+	for _, tt := range []*test{
+		{
+			name:  "Valid AES256SHA512 input", // new succeeds; doesn't attempt legacy
+			input: []byte("\xd9\x1c\x3c\x05\xb2\xf3\xc5\x93\x20\x9f\x9b\x67\x43\x8c\x0c\x3d\xe0\x80\x26\x59\x2a\x20\xb2\xe5\x5e\x30\xd6\xd1\x24\x1e\x34\x36\xbe\xfb\x79\x8e\x46\xb5\x95\xce\xe0\x79\x9c\x44\x5c\xaa\x83\x26\x92\xdb\x76\x34\x33\xe0\x0e\x0e\x54\xb2\x0b\x2f\xde\x63\x53\xf6"),
+		},
+		{
+			name:    "Invalid AES256SHA512 input", // new fails, then legacy fails
+			input:   []byte("\xda\x1c\x3c\x05\xb2\xf3\xc5\x93\x20\x9f\x9b\x67\x43\x8c\x0c\x3d\xe0\x80\x26\x59\x2a\x20\xb2\xe5\x5e\x30\xd6\xd1\x24\x1e\x34\x36\xbe\xfb\x79\x8e\x46\xb5\x95\xce\xe0\x79\x9c\x44\x5c\xaa\x83\x26\x92\xdb\x76\x34\x33\xe0\x0e\x0e\x54\xb2\x0b\x2f\xde\x63\x53\xf6"),
+			wantErr: "chacha20poly1305: message authentication failed",
+		},
+		{
+			name:  "Valid XChaCha20Poly1305 input", // new fails, then legacy succeeds
+			input: []byte("\xd9\x1c\x3c\x05\xb2\xf3\xc5\x93\x20\x9f\x9b\x67\x43\x8c\x0c\x3d\x9c\x33\x5b\x16\xd6\x9a\x9c\xf2\x9c\xf6\xe9\xbd\xdd\xe3\x1d\x54\xde\x41\xa2\x99\x56\x6a\xfc\x9a\xf3\x58\x73\x03"),
+		},
+		{
+			name:    "Invalid XChaCha20Poly1305 input", // new fails, then legacy fails
+			input:   []byte("\xda\x1c\x3c\x05\xb2\xf3\xc5\x93\x20\x9f\x9b\x67\x43\x8c\x0c\x3d\x9c\x33\x5b\x16\xd6\x9a\x9c\xf2\x9c\xf6\xe9\xbd\xdd\xe3\x1d\x54\xde\x41\xa2\x99\x56\x6a\xfc\x9a\xf3\x58\x73\x03"),
+			wantErr: "chacha20poly1305: message authentication failed",
+		},
+		{
+			name:    "Invalid length input", // both fail min length check
+			input:   make([]byte, 23),
+			wantErr: "encrypted value too short",
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			controller := gomock.NewController(t)
+			defer controller.Finish()
```

Comment:
We do not use gomock in this test, so in the current implementation we don't need a gomock controller here.

However, I think we should use mocks here instead of real implementations of `AEAD` interface because:

1. In unit tests we should strive to focus on the thing we are testing and get rid of dependencies. In this case: `multi` simly allows us to try multiple openers and we should test this. In this implementation of the test we rely on something which is not relevant to the function we test. If after some vendor update `github.com/codahale/etm` or `golang.org/x/crypto/chacha20poly1305` starts returning different error **text** - our test will fail despite the fact that everything works as expected and it is not relevant to what `multi` actualy does. Let's abstract from specific implementations of `AEAD` in the test.
1. Without mocks it is not possible to reliable test that we do not call the second opener if the first one is able to open - so there is posibility for a nasty regression which test won't catch.
1. From readig test cases  and their expected results it is not clear what we are testing: feels like we are testing `AES256SHA512` and `chacha20poly1305`.  We need to write the test and test cases so it is obvous that we are testing the following:
    * First opener is able to `Open` and we do not call second opener - _there is no way to ensure this without mocks, this is what I mean in the previous point_
    * First opener is unable to to `Open`, but second is able to `Open`
    * None of the openers are able to open - error from which opener do we return?

To be able to use mocks, we need to generate a mock for `AEAD` interface as it seems that we do not currently have it.

Reply:
I'll switch to mocks

Reply:
Unresolving this thread. I left few comments.

I was thinking about something like this:

```go
func TestOpen(t *testing.T) {
	mockInput := []byte("fakeInput")

	type test struct {
		name       string
		mocks      func(firstOpener *mock_encryption.MockAEAD, secondOpener *mock_encryption.MockAEAD)
		wantResult []byte
		wantErr    string
	}

	for _, tt := range []*test{
		{
			name: "first opener succeeds, do not try second",
			mocks: func(firstOpener *mock_encryption.MockAEAD, secondOpener *mock_encryption.MockAEAD) {
				firstOpener.EXPECT().Open(mockInput).Return([]byte("result from the first opener"), nil)
			},
			wantResult: []byte("result from the first opener"),
		},
		{
			name: "first opener errors, but second succeeds",
			mocks: func(firstOpener *mock_encryption.MockAEAD, secondOpener *mock_encryption.MockAEAD) {
				firstOpener.EXPECT().Open(mockInput).Return(nil, errors.New("fake error from the first opener"))
				secondOpener.EXPECT().Open(mockInput).Return([]byte("result from the second opener"), nil)
			},
			wantResult: []byte("result from the second opener"),
		},
		{
			name: "all openers error",
			mocks: func(firstOpener *mock_encryption.MockAEAD, secondOpener *mock_encryption.MockAEAD) {
				firstOpener.EXPECT().Open(mockInput).Return(nil, errors.New("fake error from the first opener"))
				secondOpener.EXPECT().Open(mockInput).Return(nil, errors.New("fake error from the second opener"))
			},
			wantErr: "fake error from the second opener",
		},
	} {
		t.Run(tt.name, func(t *testing.T) {
			controller := gomock.NewController(t)
			defer controller.Finish()

			firstOpener := mock_encryption.NewMockAEAD(controller)
			secondOpener := mock_encryption.NewMockAEAD(controller)

			multi := multi{
				openers: []AEAD{
					firstOpener,
					secondOpener,
				},
			}

			tt.mocks(firstOpener, secondOpener)

			b, err := multi.Open(mockInput)
			if err != nil && err.Error() != tt.wantErr ||
				err == nil && tt.wantErr != "" {
				t.Error(err)
			}
			if b != nil && !reflect.DeepEqual(tt.wantResult, b) ||
				b == nil && tt.wantResult != nil {
				t.Error(b)
			}
		})
	}
}
```